### PR TITLE
Move leading trivia with node when removing unused values.

### LINKED ISSUE #33312: "Remove redundant assignment" code fix moves variables above prepended trivia
``` C#
    class Program
    {
        static void Main(string[] args)
        {
            // Each of the last two parameters - itemType and itemName - could be null or non-null. This gives rise
            // to four such combinations. All these four combinations are valid for the project manifest but only three of them
            // are valid for property sheets.
            string foo = null; // Remove redundant assignment

            if (args.Length == 0)
            {
                foo = "1";
            }
            else
            {
                foo = "2";
            }

            Console.WriteLine(foo);
        }
    }
```

Expected:
``` C#
    class Program
    {
        static void Main(string[] args)
        {
            // Each of the last two parameters - itemType and itemName - could be null or non-null. This gives rise
            // to four such combinations. All these four combinations are valid for the project manifest but only three of them
            // are valid for property sheets.
            string foo;
            [...]
        }
    }
```

Actual:
``` C#
    class Program
    {
        static void Main(string[] args)
        {
            string foo;
            // Each of the last two parameters - itemType and itemName - could be null or non-null. This gives rise
            // to four such combinations. All these four combinations are valid for the project manifest but only three of them
            // are valid for property sheets.

            [...]
        }
    }
```

### PR DESCRIPTION
**Pull request by dibarbet**
Resolves #33312

### CODE REVIEW

File: src/EditorFeatures/CSharpTest/RemoveUnusedParametersAndValues/RemoveUnusedValueAssignmentTests.cs [hunk 1]
```diff
@@ -5401,25 +5401,24 @@ public C()
...
 
         // Unused non-constant 'out' assignment
         // Not fixed as we have a different code fix 'Use discard' for it.
         M2(out x);
-
-        // Unused initialization with only def/use in nested block.
-        // Declaration for 'y' should be moved inside the if block.
         if (flag)
         {
+            // Unused initialization with only def/use in nested block.
```
Comment (dibarbet) [inline src/EditorFeatures/CSharpTest/RemoveUnusedParametersAndValues/RemoveUnusedValueAssignmentTests.cs]: This now actually matches how the merge declaration and assignment code fix works.

If you have this
```csharp
// Trivia
int x;
if (true)
{
    x = 1;
}
```
and you merge declaration and assignment you get
```csharp
if (true)
{
    // Trivia
    int x = 1;
}
```

File: src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedValuesCodeFixProvider.cs [hunk 2]
```diff
@@ -546,7 +555,23 @@ protected sealed override async Task FixAllAsync(Document document, ImmutableArr
 
                 foreach (var node in nodesToRemove)
                 {
-                    editor.RemoveNode(node, SyntaxGenerator.DefaultRemoveOptions | SyntaxRemoveOptions.KeepLeadingTrivia);
+                    var removeOptions = SyntaxGenerator.DefaultRemoveOptions;
```
Comment (mavasani) [inline src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedValuesCodeFixProvider.cs]: I struggled a bit getting the right combination for the remove options, glad you got this working!
I wonder if this also addresses the pending trivia related issues mentioned in https://github.com/dotnet/roslyn/issues/31398#issue-384960358. If so, you should close that issue as well. Thanks!

Comment (dibarbet) [reply src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedValuesCodeFixProvider.cs]: Sure!  I'll go through the pending ones.  If any are fixed I'll update this PR with test cases that confirm the behavior.

File: src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedValuesCodeFixProvider.cs [hunk 3]
```diff
@@ -417,7 +418,7 @@ protected sealed override async Task FixAllAsync(Document document, ImmutableArr
                             // For example, "int a = 0;"
                             var variableDeclarator = node.FirstAncestorOrSelf<TVariableDeclaratorSyntax>();
                             Debug.Assert(variableDeclarator != null);
-                            nodesToRemove.Add(variableDeclarator);
+                            nodesToRemove[variableDeclarator] = false;
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedValuesCodeFixProvider.cs]: needs explanation of what true/false means.

File: src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedValuesCodeFixProvider.cs [hunk 4]
```diff
@@ -391,7 +391,8 @@ protected sealed override async Task FixAllAsync(Document document, ImmutableArr
             // at declaration, then we also add a new variable declaration statement without initializer for this local.
 
             var nodeReplacementMap = PooledDictionary<SyntaxNode, SyntaxNode>.GetInstance();
-            var nodesToRemove = PooledHashSet<SyntaxNode>.GetInstance();
+            var nodesToRemove = PooledDictionary<SyntaxNode, bool>.GetInstance();
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedValuesCodeFixProvider.cs]: this variable shoudl be renamd to better indicate what the value indicates.  i.e. nodeToXXXMapping.

File: src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedValuesCodeFixProvider.cs [hunk 5]
```diff
@@ -391,7 +391,8 @@ protected sealed override async Task FixAllAsync(Document document, ImmutableArr
...
+            var nodesToAdd = PooledHashSet<(TLocalDeclarationStatementSyntax, SyntaxNode)>.GetInstance();
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedValuesCodeFixProvider.cs]: give these tuple elements names.  seeing Item1/Item2 later isn't clear at all.

File: src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedValuesCodeFixProvider.cs [hunk 6]
```diff
@@ -594,6 +620,13 @@ void InsertLocalDeclarationStatement(TLocalDeclarationStatementSyntax declaratio
                 }
                 else if (insertionNode is TStatementSyntax)
                 {
+                    // If the insertion node is being removed, keep the leading trivia with the new declaration.
+                    if (nodesToRemove.TryGetValue(insertionNode, out var isProcessed) && !isProcessed)
+                    {
+                        declarationStatement = declarationStatement.WithLeadingTrivia(insertionNode.GetLeadingTrivia());
+                        // Mark the node as processed so that the trivia only gets added once.
+                        nodesToRemove[insertionNode] = true;
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedValuesCodeFixProvider.cs]: i'd much rather just have a second set that indicates if the node was processed or not.  then there's no need for this less clear map of nodes to booleans :)

Comment (dibarbet) [reply src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedValuesCodeFixProvider.cs]: Works for me, I'll do it that way.

File: src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedValuesCodeFixProvider.cs (hunk 2 above)
Comment (dibarbet) [reply src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedValuesCodeFixProvider.cs]: It looks like these are caused by a slightly different problem - the node isn't getting immediately removed, but an add -> replace -> remove combination.  It is similar to an issue I encountered before, so I'll see if it's easy to fix in this PR.

**Comment by sharwell** [comment]
The release build failed for #33785; re-running.

File: src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedValuesCodeFixProvider.cs (hunk 2 above)
Comment (dibarbet) [reply src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedValuesCodeFixProvider.cs]: In those other issues, it looks like the original new lines are getting replaced by elastic new lines somehow.  I've spent a few hours trying to figure out how they're getting replaced but haven't made much progress.

For now I'll merge at least this fix and look at the others if I have time later (or I find some inspiration).

Comment (mavasani) [reply src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedValuesCodeFixProvider.cs]: Sounds good to me. Thanks!

**Pull request merged**